bin/oneoff/import_concept_difficulty_tags.rb

```diff
@@ -5,37 +5,28 @@
 require 'csv'
 
 # Import level concept difficulty tags from a gsheet (downloaded as a series of
-# CSVs). When we cloned the CSF courses from 2017 to 2018, the LCDs did not get
+# CSVs). When we cloned the CSF courses from 2017 to 2018 and 2018 to 2019, the LCDs did not get
 # cloned with them so we're doing a mass import.
 
-# The given CSV contained headers that didn't map exactly to our defined LCDs,
-# so we perform a simple mapping in code
-HEADERS_TO_CONCEPTS = {
-  sequence_algorithms: ConceptDifficulties::SEQUENCING,
-  repeat: ConceptDifficulties::REPEAT_LOOPS,
-  functions_wparams: ConceptDifficulties::FUNCTIONS_WITH_PARAMS
-}
-
 def main(csv_dir)
-  %w(CourseA CourseB CourseC CourseD CourseE CourseF).each do |course|
+  %w(CourseA CourseB CourseC CourseD CourseE CourseF PreExpress Express).each do |course|
     CSV.foreach("#{csv_dir}/#{course}.csv", headers: true,  header_converters: :symbol) do |row|
-      concept_difficulties = Hash[row.to_hash.compact.map do |key, val|
-        [HEADERS_TO_CONCEPTS.fetch(key, key.to_s), val]
-      end].delete_if do |key, _val|
-        !LevelConceptDifficulty::CONCEPTS.include?(key)
-      end
-
       level = Level.find_by(name: row.fetch(:level_name))
       unless level
         puts "cannot find #{row.fetch(:level_name)}"
         next
       end
 
+      concept_difficulties = row.to_hash.compact.delete_if do |key, _val|
+        !LevelConceptDifficulty::CONCEPTS.include?(key.to_s)
+      end
+
       lcd = LevelConceptDifficulty.find_or_create_by(level: level)
 
       if concept_difficulties.any? {|key, value| lcd[key] != value}
         lcd.update!(concept_difficulties)
-        level.write_custom_level_file
+        file_path = Level.level_file_path(level.name)
+        File.write(file_path, level.to_xml)
```

Comment:
important note about this change: since we are saving the LevelConceptDifficulty record on line 27, we need to manually write to the level file. this is because:

- `level.write_custom_level_file` only writes if `should_write_custom_level_file?` is true ([here](https://github.com/code-dot-org/code-dot-org/blob/14287bb0d9a8a77025dfae51631a5f2e0bc52429/dashboard/app/models/levels/level.rb#L253)), but it is always false because...
- `level_concept_difficulty.changed?` is always false in `should_write_custom_level_file?` ([here](https://github.com/code-dot-org/code-dot-org/blob/14287bb0d9a8a77025dfae51631a5f2e0bc52429/dashboard/app/models/levels/level.rb#L248))

it felt reasonable to write directly to the level file here because we know we have just changed the LCD, but let me know if you have any feelings about this!

Reply:
Would it work to instead call `lcd.assign_attributes(concept_difficulties)` on line 27, then `level.save` here? If not, this approach is probably fine.

Reply:
unfortunately, it looks like that doesn't work either, but for a slightly different (and confusing) reason -- `level_concept_difficulty.changed?` is still false in `should_write_custom_level_file?`, but it's because `level_concept_difficulty` no longer has the newly-assigned attributes (almost like `level.reload` or `level.level_concept_difficulty.reload` was called at some point, or the lcd was re-fetched from the db?)

agreed that this is probably fine, so i'll leave as-is for now, but i was not expecting this piece to make me question my knowledge of rails 😂

Reply:
hah, gross. I'll bet it's something like the level is lazy-loading the LCD, so even though we update the one instance of lcd that we get from the database on line 24, the instance that level checks internally is a new one.